Users/binah/access-control/access-control.cs

```diff
@@ -0,0 +1,320 @@
+        {
+            if (DebugLogging) Log.Write("Script not running on an object with a physics volume!");
+        } else
+        {
+            if (DebugLogging) Log.Write("RigidBody found!");
+        }
+
+        // Subscribe to Add User events
+        // ---------------------------------------------->start coroutine TrackUser send it UserData
+        ScenePrivate.User.Subscribe(User.AddUser, SessionId.Invalid, (UserData data) => StartCoroutine(TrackUser, data.User), true);
+        if (DebugLogging) Log.Write("Subscribed to TrackUser");
+
+        //Add admins, scene owner is automatically added
+        Admins.AddRange(AdminHandles.Trim().ToLower().Split(new char[] { ',', ' ' }, StringSplitOptions.RemoveEmptyEntries));
+        Admins.Add(ScenePrivate.SceneInfo.AvatarId);
+
+        if(Admins.Count == 1)
+        {
+            //failsafe to not lock visitors into a cube when no admin is around
+            DoorsOpen = true;
+        }
+
+        // listen for commands
+        ScenePrivate.Chat.Subscribe(0, null, ShowLogCommand);
+        ScenePrivate.Chat.Subscribe(0, null, ToggleDoors);
+        ScenePrivate.Chat.Subscribe(0, null, OnBanCommand);
+        ScenePrivate.Chat.Subscribe(0, null, OnUnBanCommand);
+    }
+
+
+    public class Visitor
+    {
+        public string Name { get; internal set; }
+        public string Handle { get; internal set; }
+        public AgentPrivate Agent { get; internal set;  }
+        public bool isAdmin { get; internal set; }
+        public bool isBanned { get; internal set; }
+    }
+
+    private Dictionary<string, Visitor> Visitors = new Dictionary<string, Visitor>();
+
+    bool IsAdmin(AgentPrivate agent)
+    {
+        if (DebugLogging) Log.Write(agent.AgentInfo.Handle + " IsAdmin: " + Admins.Contains(agent.AgentInfo.Handle.ToLower()));
+        return agent != null && Admins.Contains(agent.AgentInfo.Handle.ToLower());
+    }
+
+    bool IsBanned(AgentPrivate agent)
+    {
+        if (DebugLogging) Log.Write(agent.AgentInfo.Handle + " IsBanned: " + Banned.Contains(agent.AgentInfo.Handle.ToLower()));
+        return Banned.Contains(agent.AgentInfo.Handle.ToLower());
+    }
+
+    private void TrackUser(SessionId userId)
+    {
+        AgentPrivate agent = ScenePrivate.FindAgent(userId);
+        string handle = agent.AgentInfo.Handle.ToLower();
+        Visitor visitor;
+
+        if (DebugLogging) Log.Write(handle, " has entered.");
+
+        //if visitor is not found in list add them to the list
+        if (Visitors.TryGetValue(handle, out visitor))
+        {
+            if (DebugLogging) Log.Write("Visitor entry has been found in log.");
+        } else
+        {
+            //log user entry
+            LogEntry(agent);
+        }
+
+        if (!IsBanned(agent))
+        {
+            if(DoorsOpen || IsAdmin(agent))
+            {
+                SetAccess(agent);
+            } else
+            {
+                StartCoroutine(AskForEntry, agent);
```

Comment:
this does not need to be a coroutine, you need those only to run things in parallel, usually with `while (true)` stuff

Reply:
oh okay I was following an example from leslie where had commented that modal dialogs are best done as coroutines ...